# Spatial Transcriptional Landscape of Human Heart Failure

Sang Eun Lee1,2,8, Hee Sang Hwang3,8, Shang-Fu Chen2, Douglas Evans2, Kyoung Yul Lee4, Kyung-Hee Kim5, Junho Hyun1, Min-Seok Kim1, Sung-Ho Jung6, Jae-Joong Kim1, Ali Torkamani2,7* 
###
1 Department of Cardiology, Asan Medical Center, University of Ulsan College of Medicine, Seoul, Korea
2 Scripps Research Translational Institute, La Jolla, California, USA 
3 Department of Pathology, Asan Medical Center, University of Ulsan College of Medicine, Seoul, Korea
4 Pathology Center, Seegene Medical Foundation, Seoul, Korea
5 Division of Cardiology, Cardiovascular Center, Incheon Sejong Hospital, Incheon, Korea
6 Department of Thoracic and Cardiovascular Surgery, Asan Medical Center, University of Ulsan College of Medicine, Seoul, Korea
7 Department of Integrative Structural and Computational Biology, Scripps Research, La Jolla, California, USA 

## *Correspondence: atorkama@scripps.edu

Using the GeoMx whole human transcriptome digital spatial profiler, we spatially resolved gene expression profiles in failing human hearts, accounting for cell type and clinical and histological features. 
We first described how tissue segmentation using cell-type specific antibodies could reveal unique gene expression signatures of cardiomyocytes, endothelial cells, and fibroblasts in cardiac tissue. Second, gene regulation in HF appears to be cell-type specific and may even include signatures in opposing directions across cell types, which highlights the limitation of previous bulk transcriptional profiling. Third, spatial transcriptional profiling enabled us to find detailed gene expression signatures for typical histologic features related to HF. Lastly, additional consideration of histologic features in gene expression profiling on top of clinical information provides a unique resource for further research into the tipping point for irreversible structural changes in HF progression. 

### Codes
Notebooks_Scripts_final

### Input data
"Will be uploaded soon"
